### PR TITLE
CollectSet supports structs

### PR DESCRIPTION
**Pull request by wbo4958**
This is to fix https://github.com/NVIDIA/spark-rapids/issues/3514 and https://github.com/NVIDIA/spark-rapids/issues/3152. Currently, with this PR, the CollectSet on structs can run on GPU. Since CPU and GPU can't ensure the order of collect-set, we need to add sort_array on collect-set.  But for now, sort-array didn't support on nested type, I just xfail the newly added collect-set test on structs.

### CODE REVIEW

**Comment by wbo4958** [comment]
build

**Comment by razajafri** [comment]
We should test this on DB as well?

File: integration_tests/src/main/python/hash_aggregate_test.py [hunk 1]
```diff
@@ -444,6 +450,18 @@ def test_hash_groupby_collect_set(data_gen):
             .groupby('a')
             .agg(f.sort_array(f.collect_set('b')), f.count('b')))
 
+@approximate_float
+@ignore_order(local=True)
+@incompat
+@pytest.mark.parametrize('data_gen', _gen_data_for_collect_set_op, ids=idfn)
+@pytest.mark.xfail(reason="the result order from collect-set can not be ensured for CPU and GPU."
```
Comment (revans2) [inline integration_tests/src/main/python/hash_aggregate_test.py:457]: Is there a follow on issue for this in particular?  I personally would prefer a skip over an xfail, or is there a way that we can do a simple small test where there will only ever be a single value in the set so we can at least test that case.

**Comment by wbo4958** [comment]
build

File: integration_tests/src/main/python/hash_aggregate_test.py (hunk 1 above)
Comment (wbo4958) [reply integration_tests/src/main/python/hash_aggregate_test.py:463]: Done, the new commit has added a new test case with unique group by key for this.

**Comment by wbo4958** [comment]
build

**Comment by wbo4958** [comment]
> We should test this on DB as well?

Done

**Comment by sameerz** [comment]
Please retarget to branch-21.12.

File: integration_tests/src/main/python/hash_aggregate_test.py (hunk 1 above)
Comment (wbo4958) [reply integration_tests/src/main/python/hash_aggregate_test.py:463]: @revans2 Could you help to review it again? Thx

**Review by revans2 — APPROVED** [review]

**Pull request merged**